pkg/planner/core/logical_plan_builder.go

```diff
@@ -5256,86 +5258,209 @@ func buildColumns2HandleWithWrtiableColumns(
 		}
 	}
-	sort.Sort(cols2PosInfos)
-	return cols2PosInfos, nil
+	// Sort by start position. To do the later column pruning.
+	slices.SortFunc(cols2PosInfos, func(i, j TblColPosInfo) int {
+		if i.Start < j.Start {
+			return -1
+		}
+		if i.Start > j.Start {
+			return 1
+		}
+		return 0
+	})
+	prunedColCnt := 0
+	var err error
+	for i := range cols2PosInfos {
+		cols2PosInfo := &cols2PosInfos[i]
+		tbl := tblID2Table[cols2PosInfo.TblID]
+		tblInfo := tbl.Meta()
+		// If it's partitioned table, or has foreign keys, or is point get plan, we can't prune the columns, currently.
+		// nonPrunedSet will be nil if it's a point get or has foreign keys.
+		if tblInfo.GetPartitionInfo() != nil || hasFK || nonPruned == nil {
+			err = buildSingleTableColPosInfoForDelete(tbl, cols2PosInfo)
+			if err != nil {
+				return nil, nil, err
+			}
+			continue
+		}
+		prunedColCnt, err = pruneAndBuildSingleTableColPosInfoForDelete(tbl, tblInfo, names, cols2PosInfo, prunedColCnt, nonPruned)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return cols2PosInfos, nonPruned, nil
+}
+
+// initColPosInfo initializes the column position information.
+// It's used before we call pruneAndBuildSingleTableColPosInfoForDelete or buildSingleTableColPosInfoForDelete.
+// It initializes the needed information for the following pruning: the tid, starting position and the handle column.
+func initColPosInfo(tid int64, names []*types.FieldName, handleCol util.HandleCols) (TblColPosInfo, error) {
+	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
+	if err != nil {
+		return TblColPosInfo{}, err
+	}
+	return TblColPosInfo{
+		TblID:                 tid,
+		Start:                 offset,
+		HandleCols:            handleCol,
+		ExtraPartialRowOption: &table.ExtraPartialRowOption{},
+	}, nil
 }
 
-// buildSingleTableColPosInfoForDelete builds columns mapping for delete.
+// buildSingleTableColPosInfoForDelete builds columns mapping for delete without pruning any columns.
+// It's temp code path for partition table, foreign key and point get plan.
 func buildSingleTableColPosInfoForDelete(
-	names []*types.FieldName,
-	handleCol util.HandleCols,
-	deletableIdxs []table.Index,
-	deletableCols []*table.Column,
+	tbl table.Table,
+	colPosInfo *TblColPosInfo,
+) error {
+	deletableIdxs := tbl.DeletableIndices()
+	tblLen := len(tbl.DeletableCols())
+	indexColMap := make(map[int64]table.IndexRowLayoutOption, len(deletableIdxs))
+	for _, idx := range deletableIdxs {
+		idxCols := idx.Meta().Columns
+		colPos := make([]int, 0, len(idxCols))
+		for _, col := range idxCols {
+			colPos = append(colPos, col.Offset)
+		}
+		indexColMap[idx.Meta().ID] = colPos
+	}
+	colPosInfo.End = colPosInfo.Start + tblLen
+	colPosInfo.IndexesRowLayout = indexColMap
+	return nil
+}
+
+// pruneAndBuildSingleTableColPosInfoForDelete builds columns mapping for delete.
+// And it will try to prune columns that not used in pk or indexes.
+func pruneAndBuildSingleTableColPosInfoForDelete(
```

Comment:
This function is too large, unnecessarily complicated, and hard to understand.
We should initialize `TblColPosInfo`, do pruning, and collect information for the tables one by one, instead of initializing them as if there's no pruning, and then "fix" everything when we prune them.

Reply:
The current implementation achieves the optimal time complexity.